cumulus/client/cirrus-executor/src/bundle_processor.rs

```diff
@@ -513,6 +518,42 @@ where
 		Ok(())
 	}
 
+	fn try_submit_fraud_proof_for_first_unconfirmed_bad_receipt(
+		&self,
+		oldest_receipt_number: NumberFor<PBlock>,
+	) -> Result<(), sp_blockchain::Error> {
+		if let Some((bad_receipt_number, _bad_signed_receipt_hash, _trace_mismatch_index)) =
+			crate::aux_schema::load_first_unconfirmed_bad_receipt_info::<_, NumberFor<PBlock>>(
+				&*self.client,
+				oldest_receipt_number,
+			)? {
+			let block_number: BlockNumber = bad_receipt_number
+				.try_into()
+				.unwrap_or_else(|_| panic!("Primary number must fit into u32; qed"));
+			let block_hash = self.client.hash(block_number.into())?.ok_or_else(|| {
+				sp_blockchain::Error::Backend(format!(
+					"Header hash not found for number {block_number}"
+				))
+			})?;
+			let _local_receipt = crate::aux_schema::load_execution_receipt::<
+				_,
+				Block::Hash,
+				NumberFor<PBlock>,
+				PBlock::Hash,
+			>(&*self.client, block_hash)?
+			.ok_or_else(|| {
+				sp_blockchain::Error::Backend(format!(
+					"Execution receipt not found for {block_hash:?}",
+				))
+			})?;
```

Comment:
It is hard for me to follow across the whole thing, but isn't it possible that there are 2 blocks at the same height here and referencing should be done by block hash in the first place?

Reply:
Yeah, I get your point, it's indeed problematic considering the forks that the receipt used to generate the fraud proof later can be not the same as the one `trace_mismatch_index` was generated from before. That being said, we need to also store the related `local_receipt_hash` along with the `trace_mismatch_index` and ensure `local_receipt_hash` is not on a fork when preparing the fraud proof.

Reply:
So are you going to apply some tweaks to handle this?

Reply:
Yeah, I'm working on that, just more changes than I thought initially.

Reply:
https://github.com/subspace/subspace/pull/652/commits/5129cf78e3caa72f2b65227f50b46be8cca4b296 Some renamings about the storage key are put into the next commit to reduce some code changes for easier review. It's possible to simply find the first unconfirmed bad receipt on the canonical chain, but given it's not clear how we'll process the fork receipt info later, I just leave it as it is for now.